File: apps/mybookkeeper/frontend/src/shared/types/utility/utility-plan-create-request.ts
import type { UtilityPlanRateType } from "@/shared/types/utility/utility-plan-rate-type";
import type { UtilityServiceType } from "@/shared/types/utility/utility-service-type";

/**
 * Body for ``POST /utility-plans``.
 *
 * Mirrors ``schemas/properties/utility_plan_create_request.py``, which sets
 * ``extra="forbid"`` — an unknown key is a 422, not a silently dropped field.
 *
 * Rate fields are sent as strings so sub-cent precision survives the JSON
 * boundary: a TDU charge of 5.3509 ¢/kWh rounded to 5.35 by a float round-trip
 * would misprice every comparison built on it.
 */
export interface UtilityPlanCreateRequest {
  property_id: string;
  service_type: UtilityServiceType;
  provider_name: string;
  rate_type: UtilityPlanRateType;
  account_number?: string | null;
  plan_name?: string | null;

  energy_charge_cents_per_kwh?: string | null;
  tdu_charge_cents_per_kwh?: string | null;
  avg_price_cents_per_kwh_at_1000?: string | null;
  monthly_base_charge_cents?: number | null;

  term_months?: number | null;
  service_start_date?: string | null;
  term_end_date?: string | null;
  early_termination_fee_cents?: number | null;

  has_bill_credit?: boolean;
  bill_credit_amount_cents?: number | null;
  bill_credit_threshold_kwh?: number | null;
  min_usage_fee_cents?: number | null;
  min_usage_threshold_kwh?: number | null;

  notes?: string | null;
}

File: apps/mybookkeeper/backend/app/schemas/properties/utility_plan_create_request.py
"""Schema for POST /utility-plans."""
from __future__ import annotations

import datetime as _dt
import uuid
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.schemas.properties.utility_plan_validation import (
    MAX_RATE_CENTS,
    validate_plan_fields,
)


class UtilityPlanCreateRequest(BaseModel):
    property_id: uuid.UUID
    service_type: str
    provider_name: str = Field(..., min_length=1, max_length=255)
    rate_type: str
    account_number: str | None = Field(None, max_length=100)
    plan_name: str | None = Field(None, max_length=255)

    energy_charge_cents_per_kwh: Decimal | None = Field(
        None, ge=0, le=MAX_RATE_CENTS, decimal_places=4,
    )
    tdu_charge_cents_per_kwh: Decimal | None = Field(
        None, ge=0, le=MAX_RATE_CENTS, decimal_places=4,
    )
    avg_price_cents_per_kwh_at_1000: Decimal | None = Field(
        None, ge=0, le=MAX_RATE_CENTS, decimal_places=4,
    )
    monthly_base_charge_cents: int | None = Field(None, ge=0)

    term_months: int | None = Field(None, ge=0, le=600)
    service_start_date: _dt.date | None = None
    term_end_date: _dt.date | None = None
    early_termination_fee_cents: int | None = Field(None, ge=0)

    has_bill_credit: bool = False
    bill_credit_amount_cents: int | None = Field(None, ge=0)
    bill_credit_threshold_kwh: int | None = Field(None, ge=0)
    min_usage_fee_cents: int | None = Field(None, ge=0)
    min_usage_threshold_kwh: int | None = Field(None, ge=0)

    notes: str | None = Field(None, max_length=5000)

    model_config = ConfigDict(extra="forbid")

    @model_validator(mode="after")
    def _check_fields(self) -> UtilityPlanCreateRequest:
        return validate_plan_fields(self)
